[DROOLS-3570]  Add validation of simulation structure

danielezonca opened the pull request:
Implemented validation of structure of test scenario.

These are check on RULE test scenario:
-> empty column skip
-> instance type removed
-> navigation of bean still valid
-> field type changed

These are check on DMN test scenario
-> empty column skip
-> DMN node removed
-> simple type becomes complex type
-> navigation of data type still valid
-> field type changed

Validation is performed during the loading of the editor or via `Validate` button

https://issues.jboss.org/browse/DROOLS-3570

@kkufova @gitgabrio @Rikkola @yesamer

gitgabrio commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-backend/src/main/java/org/drools/workbench/screens/scenariosimulation/backend/server/ScenarioValidationService.java:
```diff
@@ -0,0 +1,244 @@
+                factMapping.getExpressionElements().size() == 0;
+    }
+
+    private DMNType navigateDMNType(DMNType rootType, List<String> steps) {
```
@danielezonca 
I think it would be better to move this and the following method to a "utility" class - DMNTypeUtils for the moment being has only one method, but if it would be refactored as a common "DMNUtil", it would be the best place to put all this DMN-related helper methods

danielezonca replied:
Done

gitgabrio commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/editor/ScenarioSimulationEditorPresenter.java:
```diff
@@ -470,6 +499,8 @@ public void getModelSuccessCallbackMethod(DataManagementStrategy dataManagementS
         context.getStatus().setSimulation(model.getSimulation());
         CustomBusyPopup.close();
+        // check if structure is valid
+        getValidateCommand().execute();
```
@danielezonca 
I think this specific back-and-forth may and should be avoided. We can add those "FactMappingError" inside "ScenarioSimulationModel", mark them to be ignored by Xstream (=not persisted on xml) and showing them at load, without having to issue another service call. Nothing would prevent, anyway, to let the user call the "validate" service at will.

danielezonca replied:
I prefer to postpone this refactor so I have opened this ticket to track it
https://issues.jboss.org/browse/DROOLS-4435

danielezonca commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-backend/src/main/java/org/drools/workbench/screens/scenariosimulation/backend/server/ScenarioValidationService.java:
```diff
@@ -0,0 +1,244 @@
+
+            DMNType rootDMNType;
+            try {
+                rootDMNType = dmnModel.getDecisionByName(nodeName) != null ?
```
@gitgabrio 
No, the behavior is correct because user can edit the template after the creation and move a decision from EXPECT to GIVEN to test another decision and the opposite

danielezonca commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-backend/src/main/java/org/drools/workbench/screens/scenariosimulation/backend/server/ScenarioValidationService.java:
```diff
@@ -0,0 +1,244 @@
+import static org.kie.dmn.feel.lang.types.BuiltInType.UNKNOWN;
+
+@ApplicationScoped
+public class ScenarioValidationService
```
Done

kkufova commented:
@danielezonca, so far, I found the following issues:

1. In rule-based test scenarios, if you remove a data object that is used in the test scenario table, the designer correctly shows the **Wrong column mapping** dialog and Business Central freezes afterwards. Moreover, if this data object is still used inside a DRL file, after clicking **Validate**, an error is thrown instead:

```
Unable to complete your request. The following exception occurred: java.lang.IllegalArgumentException:Retrieving KieContainer has failed. Fix all compilation errors within the project and build the project again..
```

2. In general, my problem with the validation is that it happens quite often that after validating, the **Loading...** message appears and Business Central freezes (and must be reloaded). It makes the testing almost impossible. I got into situation when the loading popup appears every time a test scenario asset is opened and it cannot be edited at all.

jomarko commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-backend/src/main/java/org/drools/workbench/screens/scenariosimulation/backend/server/DMNScenarioValidation.java:
```diff
@@ -0,0 +1,126 @@
+        if (Objects.equals(factMappingType, dmnType.getName())) {
+            return true;
+        }
+        return false;
```
I think [1] is same as [2]

[1]
```java
if (Objects.equals(factMappingType, dmnType.getName())) {
            return true;
        }
        return false;
```

[2]
```java
return Objects.equals(factMappingType, dmnType.getName());
```

danielezonca replied:
Yes, you are right, done :)

jomarko commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-backend/src/main/java/org/drools/workbench/screens/scenariosimulation/backend/server/AbstractScenarioValidation.java:
```diff
@@ -0,0 +1,51 @@
+    protected boolean isToSkip(FactMapping factMapping) {
+        return OTHER.equals(factMapping.getExpressionIdentifier().getType()) ||
+                EMPTY.equals(factMapping.getFactIdentifier()) ||
+                factMapping.getExpressionElements().size() == 0;
```
Just for the sake of short codebase `.size() == 0` could be replaced by `.isEmpty()`

danielezonca replied:
Done

jomarko commented on drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-backend/src/main/java/org/drools/workbench/screens/scenariosimulation/backend/server/DMNScenarioValidation.java:
```diff
@@ -0,0 +1,126 @@
+            // error if direct mapping (= simple type) but it is a composite
+            // NOTE: context is a special case so it is composite even if no fields are declared
+            Type rootType = getRootType((BaseDMNTypeImpl) rootDMNType);
+            if (!CONTEXT.equals(rootType) && steps.size() == 0 && rootDMNType.isComposite()) {
```
We could use `steps.isEmpty()` as we do in `RULEScenarioValidation`

danielezonca replied:
Done

danielezonca commented:
@jomarko 
Can you please test it?
I have fixed both issues raised by Klara:
1. A notification error appear if validation fail because the compilation of the project fails (please note that DMN test scenario has another error popup raised by right panel too in case of compilation failure and this has not been changed)
![Screenshot from 2019-08-19 18-48-32](https://user-images.githubusercontent.com/244607/63283638-10015200-c2b2-11e9-8b55-05eed813c67f.png)
2. Removed loading indicator after validation
